enhancements/insights/insights-operator-runtime.md

```diff
@@ -0,0 +1,532 @@
+    * Value is a hash of the detected runtime version: `2.10.3.Final, 7.4.7.GA, 2.7.1, 3.6.0.redhat-00005,...` (list not exhaustive)
+
+It is not planned to add more fields to the `workloadRuntimeInfoContainer` type. However, the insights-operator-runtime component can be enhanced to detect more runtimes without impacting the integration in the Insights Operator.
+
+##### Examples
+
+```json
+"runtimeInfo": {
+  "os": "yqx1WBefoIAq",
+  "osVersion": "AQN-qh_OyRUA",
+  "kind": "wbpgzhNYZQOi",
+  "kindVersion": "cAsoUDy8JFNV",
+  "kindImplementer": "DzP5-GS6tDOQ",
+  "runtimes": [
+    {
+      "name": "zBVuVhrrC_vC",
+      "version": "lyAE7Oh63WPU"
+    }
+  ]
+}
+```
+
+As the values are derived by the insights-operator-runtime component or correspond to software versions, it is possible to create a dictionary of the hashes and retrieve the unobfuscated values. The example above corresponds to:
+
+```json
+"runtimeInfo": {
+  "os": "rhel",
+  "osVersion": "8.9",
+  "kind": "Java",
+  "kindVersion": "17.0.7",
+  "kindImplementer": "Red Hat, Inc.",
+  "runtimes": [
+    {
+      "name": "Quarkus",
+      "version": "2.13.8.Final-redhat-00004"
+    }
+  ]
+}
+```
+
+### Workflow Description
+
+In its existing implementation, the OpenShift Insights Operator periodically gathers workload data every 12 hours.
+
+The Operator iterates over pods in the cluster (up to a limit of 8000) and, for each pod's containers, captures their Image ID and entry points.
+
+When the operator starts gathering workload information, it would first gather runtime information for all the containers running on the cluster's worker nodes and then
+populate the existing gathered info with these additional runtime information.
+
+![insights-operator-runtime workflow](./insights-operator-runtime-1.png "Insights Operator Runtime Worklow")
+
+This is not a user-facing feature. It is achieved within the OpenShift Insights Operator gathering workflow.
+The user is able to examine the gathered data by unarchiving the archive generated by the operator.
+
+The DaemonSet for the insights-operator-runtime is deployed when the Insights Operator is installed in the cluster and is not accessible outside of the `openshift-insights` namespace.
+
+From the operator perspective, the insights-operator-runtime is a black box, and its requirements are:
+
+* Determine the Pod name of each insights-operator-runtime’s Pod (as the insights-operator-runtime is deployed as a DaemonSet, there is a single pod instance on each worker node) and map it to its worker node.
+  * There is no service exposing the insights-operator-runtime and its pods are accessed directly.
+* When the operator starts gathering workload information, execute the  `/scan-containers` executable on each insights-operator-runtime pod. The executable will return a JSON payload in the standard output that represents a tree of `namespace/pod-name/container-id/runtime-info`.
```

Comment:
I'm not clear on why the API here is "operator executes a `/scan-containers` call in each DaemonSet pod" instead of a [kube-rbac-proxy][1]-gated HTTP endpoint, or something else that allows for the long-running DaemonSet pod to do something more service-y in the future.  Can you add an *Alternatives* subsection about this, laying out some pros and cons?  My guess is that executing `/scan-containers` calls is easier for the insights-operator-runtime to implement, but maybe a bit harder for the Insights operator to call into, and there's some expectation that insights-operator-runtime-dev-capacity will be tighter than Insights-operator-dev-capacity?

[1]: https://github.com/openshift/kube-rbac-proxy

Reply:
> I'm not clear on why the API here is "operator executes a /scan-containers call in each DaemonSet pod" instead of a [kube-rbac-proxy](https://github.com/openshift/kube-rbac-proxy)-gated HTTP endpoint, or something else that allows for the long-running DaemonSet pod to do something more service-y in the future.

I like this idea, but it's not clear to me how the data would be aggregated behind the endpoint :thinking:

Reply:
Prometheus has [ServiceMonitors][1] asking for scraping, and then Prometheus walks each selector-matching Pod and scrapes the configured endpoints.  The Insights operator here will be walking DaemonSet Pods, and it can do whatever on each Pod to get data out.  The current enhancement calls for getting into the Pod's container and running `/scan-containers`.  But I don't see why it couldn't call for making an HTTPS request against an endpoint served by that Pod instead ([Pod DNS record docs][2]).

[1]: https://docs.openshift.com/container-platform/4.15/observability/monitoring/managing-metrics.html#specifying-how-a-service-is-monitored_managing-metrics
[2]: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#a-aaaa-records-1

Reply:
The original intent of this RFE was that the Insights Operator would deploy the daemon set on demand, get the data from it and undeploy it. Having a HTTP server instead of a CLI executable would not have make much difference.

@tremes mentioned that the Insights Operator can now create gather jobs on demand and in parallel so I updated the RFE to have the daemon set always deployed. I can change the RFE to get the data from a HTTP server (based on the pods DNS record (thanks for the link) instead of the executable but there are a few question with that change?

* Does that make sense to have a HTTP server always up and running when it will be called a few times (every 12 hours + on demand gather jobs). OpenShift will have to monitor that the server is always up and running even if we know it won't be called by anybody but the Insights Operator during specific time window?
* kube-rbac-proxy is a project is alpha stage. Does using it affect the state of this RFE (might not be relevant for Tech Preview but could be for GA)

Reply:
> OpenShift will have to monitor that the server is always up and running even if we know it won't be called by anybody but the Insights Operator during specific time window?

Why is this?  If we don't care if it's up outside of Insights-operator scrapes, won't those scrapes provide the monitoring we want?

 > kube-rbac-proxy is a project is alpha stage. Does using it affect the state of this RFE...
 
 https://amd64.ocp.releases.ci.openshift.org/ -> [4.15.19](https://amd64.ocp.releases.ci.openshift.org/releasestream/4-stable/release/4.15.19) -> [aws-ovn-serial](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.15-e2e-aws-ovn-serial/1803885981464530944) -> [Artifacts](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.15-e2e-aws-ovn-serial/1803885981464530944/) -> ... -> [gather-extra artifacts](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.15-e2e-aws-ovn-serial/1803885981464530944/artifacts/e2e-aws-ovn-serial/gather-extra/artifacts/):
 
 ```console
$ oc adm release info --image-for kube-rbac-proxy quay.io/openshift-release-dev/ocp-release:4.15.19-x86_64
quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:f15be2dc3d6d2c1848f997b7921d11bf4c0873c129cdf182155ac133880b0a89
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.15-e2e-aws-ovn-serial/1803885981464530944/artifacts/e2e-aws-ovn-serial/gather-extra/artifacts/pods.json | jq -r '.items[] | .metadata as $m | .status.containerStatuses[] | select(.imageID | endswith("@sha256:f15be2dc3d6d2c1848f997b7921d11bf4c0873c129cdf182155ac133880b0a89")) | $m.namespace + " " + $m.name + " " + .name' | sort
openshift-cloud-controller-manager-operator cluster-cloud-controller-manager-operator-8dc747bf5-6jfbh kube-rbac-proxy
openshift-cloud-credential-operator cloud-credential-operator-6f8c5bc85-gtb6b kube-rbac-proxy
openshift-cluster-csi-drivers aws-ebs-csi-driver-controller-d5578f5db-6p6mp attacher-kube-rbac-proxy
openshift-cluster-csi-drivers aws-ebs-csi-driver-controller-d5578f5db-6p6mp kube-rbac-proxy-8201
openshift-cluster-csi-drivers aws-ebs-csi-driver-controller-d5578f5db-6p6mp provisioner-kube-rbac-proxy
openshift-cluster-csi-drivers aws-ebs-csi-driver-controller-d5578f5db-6p6mp resizer-kube-rbac-proxy
...many hits...
```

So lots of precedent for using the `kube-rbac-proxy`.  I expect the alpha-ness is related to customer use, and not internal-to-OCP-core use?  Do you have a reference for it being alpha, in case that has the "for who?" context?

Reply:
@wking The "alpha stage" is mentioned in https://github.com/openshift/kube-rbac-proxy README. If there is a precedent to use it, I've no objection. I've updated the RFE to mention its use (as well as using a HTTP server to query the runtime info)